### ISSUE TITLE
docs: explain better the distinction between islands and components

### ISSUE THREAD

**Issue by skybrian** [denoland/fresh#2369]
Quoting @CAYdenberg in: https://github.com/denoland/fresh/issues/2363#issuecomment-2004817651

> Well, if it helps, components that are children of islands can receive functions as props. So, your ComplexOrBigComponent could live in components/, receive onClick when used as a descendent of an island, and not receive it when used outside of an island. Just mark onClick as being optional in your props interface declaration (and handle the cases where it is undefined).

I also found this confusing at first. When should a file go under components/ versus islands/?

 I had thought that islands were client-side and components were server-side, but that's not right. Instead, everything should be a component, except when it needs to be the root of a client-side hierarchy, and then it's an island. (Why? Because there are restrictions on what you can pass in to an island.)

In particular, the [component page](https://fresh.deno.dev/components) has some components tagged as islands that are really just components that can run on both client and the server. (A button could be an island, but why?)

**Comment by fro-profesional** [comment]
I don think is confusing, from the docs: https://fresh.deno.dev/docs/concepts/islands

> Islands enable client side interactivity in Fresh. Islands are isolated Preact components that are rendered on the server and then hydrated on the client. This is different from all other components in Fresh, as they are usually rendered on the server only.

You mentioned

> In particular, the [component page](https://fresh.deno.dev/components) has some components tagged as islands that are really just components that can run on both client and the server.

Thats what islands are, thats what the docs says "components that are rendered on the server and then hydrated on the client."

And for

> (A button could be an island, but why?)

Why would not? buttons can be used to trigger forms by using props, or actions by onClick event, for the second will always be necessary to be an island or be used inside and island

**Comment by CAYdenberg** [comment]
I think based on the fact that multiple people are _confused_ (who presumably took the time to try to understand, since they took the time to give feedback), that by definition it is _confusing_.

I think the concept that needs to be fleshed out is that the children of islands are still "within" the island. Or, put differently, a distinction should be made between the island "entry-point" and the components which are part of the island.

However, I am not a maintainer, so I won't make the call myself. It's also possible the change should be folded into the broader re-org of the Getting Started section (#1632 )

**Comment by fro-profesional** [comment]
> I think based on the fact that multiple people are confused (who presumably took the time to try to understand, since they took the time to give feedback), that by definition it is confusing.

I guess u are right 🤔

**Comment by marvinhagemeister** [comment]
## Component
A function that starts with an uppercase letter that can use hooks and optionally return JSX. Doesn't say anything about where it runs.

## Island
A form of component that is bundled and executed it in the browser, looks the exact same as a normal component. The only difference is that they're put in special files like `islands/MyIsland.tsx` and Fresh flags components exported from these as an island.

## Server Component
A form of component that's only run on the server. Looks the exact same as a normal component, only difference is where it's used. For Fresh means that if you only ever use this component in a route, then it will only run on the server.

**Comment by kerimhudson** [comment]
Just as an anecdote, I did find this somewhat confusing at first, especially because I copied the Button island into a component in my Fresh project, and then was confused why it was constantly disabled. Of course, now this makes sense to me.

But then when I look at the interactive islands documentation now, looking at the following example:
```
islands/my-island.tsx
import { useSignal } from "@preact/signals";

export default function MyIsland() {
  const count = useSignal(0);

  return (
    <div>
      Counter is at {count}.{" "}
      <button onClick={() => (count.value += 1)}>+</button>
    </div>
  );
}
```

It's now clear to me that the button here, could be a `components/Button` and still function perfectly fine. I wonder if in a documentation level it's worth just creating an example which illustrates this use case a little more? Something like:

```
islands/my-island.tsx
import { useSignal } from "@preact/signals";
import Button from '../components/Button.tsx'

export default function MyIsland() {
  const count = useSignal(0);

  return (
    <div>
      Counter is at {count}.{" "}
      <Button onClick={() => (count.value += 1)}>+</Button>
    </div>
  );
}
```

**Comment by crowlsyong** [comment]
<sub>Note: I am a hobbyist developer. Words like: novice, mediocre, hilariously uninformed, and confused are words that describe my programming journey most of the time. If you're wondering whether or not I'm justifiabley confused, I probably am not and just need to read more docs. </sub>

**tldr i also have some confusion but it's probably my fault. i invite help.** 

My projects are hobby projects and small, so I haven't spent enough time really learning this and **I end up just defaulting to making everything an island** because I don't want to run into issues about certain things not being able to run (islands seem to always just work! haha). 

I understand that basically everything is serverside except islands which ship javascript to the client (or maybe I don't understand this as well as I think I do)

To illustrate, I'm working on an app with sliders, buttons, and text that are all wired up in different ways. See gif below: 
![Image](https://github.com/user-attachments/assets/9636a56f-887e-4a35-ad56-de043a882b8c)

**Which things in this gif should be components and which islands, or do you need more info from me to decide, or is it a little bit ambiguous in some cases, or other?**
- I have a chart from chartjs, should that js/ts be rendered server or client side (component or island)? Or do you need more info to determine this?
- I have a button that, when clicked, shifts points on the chart by performing some math, should that button be a component, island, other, or need more info?
- Should I make the entire slider row an island, or make it a component and put a tiny little island part inside of it? How much does it matter when it's 50 lines of TypeScript vs 500 vs 5,000 vs 50,000?

---

I also don't know if there is GitHub etiquette around commenting on Closed issues, which I have just done by posting this comment. Tell me if I should be commenting this elsewhere. If no one says otherwise, I'll assume this is okay to do and not annoying. 

<sub>Thanks for all the help and kindness everyone. shout out to marvin for being the goat. that fella is always kind, patient, and helpful. </sub>

**Comment by CAYdenberg** [comment]
@crowlsyong Basically, anything that changes without having to reload the page, or that triggers a change, should be an island. In certain cases they could be _separate_ islands, if you're using Signals to manage the shared state.

In your case it looks like basically the whole of the screenshot you posted, except maybe the header row. You still get some benefits from the island architecture, in the sense that the page metadata, layout, etc., would be serverside rendered only, but it's a minimal benefit in your case.

I'd also note that Chart.js renders on canvas, so your chart wouldn't render serverside or without JavaScript, anyway.

**Comment by crowlsyong** [comment]
Y'all are helpful and patient and it warms my heart. thank you everyone.
